File: leetcode/21.合并两个有序链表.java
/*
 * @lc app=leetcode.cn id=21 lang=java
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
     //迭代法
     //拟定一个头节点，根据题意可以使用最小值
     ListNode preHead = new ListNode(Integer.MIN_VALUE);
     //当前指针指向头节点
     ListNode prev = preHead;
     //比较两个链表的值，当前指针指向链表较小的节点，
     //修改链表指针指向下一个节点
     while(l1!=null&&l2!=null){
         if(l1.val<=l2.val){
             prev.next = l1;
             l1 = l1.next;
         }else{
             prev.next = l2;
             l2 = l2.next;
         }
             prev = prev.next;
     }
     //最后将当前指针指向的节点接上剩余的节点
     prev.next = l1!=null?l1:l2;
     //前置头节点是自己定义的，真正的头节点是它的下一个节点
     return preHead.next;
    }
}
// @lc code=end

File: leetcode/21.合并两个有序链表.rs
/*
 * @lc app=leetcode.cn id=21 lang=rust
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut l1 = l1;
        let mut l2 = l2;
        //头节点
        let mut head = Some(Box::new(ListNode::new(i32::MIN)));
        //当前指针
        let mut cur = &mut head;
        loop {
            match (l1.take(),l2.take()) {
                (None,Some(v))|(Some(v),None) => {
                    if let Some(ref mut p) = cur {
                        p.next = Some(v);
                    }
                    break;
                },
                //(Some(v),None) => {
                //    if let Some(ref mut p) = cur {
                //        p.next = Some(v);
                //    }
                //    break;
                //},
                (Some(mut v1),Some(mut v2)) => {
                    let mut node = None;
                   if v1.val >= v2.val {
                       //l2指向下一个，这里要take出来，
                       //因为v2要移动到Some里
                       l2 = v2.next.take();
                       node = Some(v2);
                       //l1被take了这里要还原
                       l1 = Some(v1);
                   }else{
                       //l1指向下一个，这里要take出来，
                       //因为v1要移动到Some里
                       l1 = v1.next.take();
                       node = Some(v1);
                       //l2被take了这里要还原
                       l2 = Some(v2);
                   }
                   if let Some(ref mut p) = cur {
                      p.next = node;
                      cur = &mut p.next;
                   }
                },
                (None,None) => {
                    break;
                },
            }
        }

        head.unwrap().next
    }
}
// @lc code=end

